### PR TITLE
examples/boto3/README: examples notification

### PR DESCRIPTION
**Pull request by dorindabassey**
added examples on how to fetch attributes of a specific topic

Fixes: https://tracker.ceph.com/issues/46243

Signed-off-by: dorindabassey <dorindabassey@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: examples/boto3/README.md [hunk 1]
```diff
@@ -13,12 +13,21 @@ The standard [AWS CLI](https://docs.aws.amazon.com/cli/latest/) may also be used
...
+```
 - Bucket notifications with filtering extensions:
 ```
 aws --region=default --endpoint-url http://localhost:8000 s3api put-bucket-notification-configuration --bucket mybucket --notification-configuration='{"TopicConfigurations": [{"Id": "notif1", "TopicArn": "arn:aws:sns:default::mytopic",  
 "Events": ["s3:ObjectCreated:*", "s3:ObjectRemoved:*"],  
 "Filter": {"Metadata": {"FilterRules": [{"Name": "x-amz-meta-foo", "Value": "bar"}, {"Name": "x-amz-meta-hello", "Value": "world"}]}, "Key": {"FilterRules": [{"Name": "regex", "Value": "([a-z]+)"}]}}}]}'
  ```
+- Topic creation with endpoint:
+```
+aws --endpoint-url http://localhost:8000 s3api create-topic --name=myname --attributes=myattributes
```
Comment (yuvalif) [inline examples/boto3/README.md:29]: what is "myattributes"?

File: examples/boto3/topic_attributes.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        aws_secret_access_key=secret_key,
+        region_name=region_name,
+        config=Config(signature_version='s3'))
+
+# to see the list of available "regions" use:
+# radosgw-admin realm zonegroup list
+
+# getting a list of topics is an extension to AWS S3 API
+
+print client.list_topics(TopicArn=topic_arn)
```
Comment (yuvalif) [inline examples/boto3/topic_attributes.py:37]: how is `list_topics()` related to getting the topic's attributes?
please see [here](https://github.com/ceph/ceph/blob/master/src/test/rgw/rgw_multi/zone_ps.py#L212) example for getting the topic.

File: examples/boto3/README.md [hunk 3]
```diff
@@ -13,12 +13,21 @@ The standard [AWS CLI](https://docs.aws.amazon.com/cli/latest/) may also be used
...
+```
 - Bucket notifications with filtering extensions:
 ```
 aws --region=default --endpoint-url http://localhost:8000 s3api put-bucket-notification-configuration --bucket mybucket --notification-configuration='{"TopicConfigurations": [{"Id": "notif1", "TopicArn": "arn:aws:sns:default::mytopic",  
 "Events": ["s3:ObjectCreated:*", "s3:ObjectRemoved:*"],  
 "Filter": {"Metadata": {"FilterRules": [{"Name": "x-amz-meta-foo", "Value": "bar"}, {"Name": "x-amz-meta-hello", "Value": "world"}]}, "Key": {"FilterRules": [{"Name": "regex", "Value": "([a-z]+)"}]}}}]}'
  ```
+- Topic creation with endpoint:
+```
+aws --endpoint-url http://localhost:8000 s3api create-topic --name=myname --attributes={nvp[0] : nvp[1]}
```
Comment (yuvalif) [inline examples/boto3/README.md:29]: its is not clear where you get the "nvp" array from? please test the example first.
note this example is not inside a python script. this is just CLI called from the shell.

File: examples/boto3/topic_attributes.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        aws_secret_access_key=secret_key,
+        region_name=region_name,
+        config=Config(signature_version='s3'))
+
+# to see the list of available "regions" use:
+# radosgw-admin realm zonegroup list
+
+# getting attributes of a specific topic is an extension to AWS S3 API
+
+print client.get_topic(TopicArn=topic-arn)
```
Comment (yuvalif) [inline examples/boto3/topic_attributes.py:37]: i'm afraid we support the official "get_topic()" API yet.
example here should include the "hand craftet" REST call. see [here](https://github.com/ceph/ceph/blob/master/src/test/rgw/rgw_multi/zone_ps.py#L212)
note that we are not actually using boto3 there, but a plain HTTP cleint.

**Review by yuvalif — CHANGES_REQUESTED** [review]
comments in the code

File: examples/boto3/README.md [hunk 5]
```diff
@@ -26,7 +26,7 @@ aws --region=default --endpoint-url http://localhost:8000 s3api put-bucket-notif
  ```
 - Topic creation with endpoint:
 ```
-aws --endpoint-url http://localhost:8000 s3api create-topic --name=myname --attributes={nvp[0] : nvp[1]} 
+aws --endpoint-url http://localhost:8000 s3api create-topic --name=myname --attributes={'TopicArn': 'arn:aws:sns:region:acct-id:mytopic'}
```
Comment (yuvalif) [inline examples/boto3/README.md:29]: topicarn is returned from the call. not passed to the call.
you need to pass attributes. take this: https://github.com/ceph/ceph/blob/master/examples/boto3/topic_with_endpoint.py#L38
is an example, but pass the attributes explictly (not via the URL-like parsing), just do:
```
--attributes{'push-endpoint': 'amqp://127.0.0.1:5672', 'amqp-exchange': 'ex1', 'amqp-ack-level': 'broker'}
```
and please test before you update the doc.

Comment (dorindabassey) [reply examples/boto3/README.md:29]: Alright thanks.

File: examples/boto3/README.md [hunk 6]
```diff
@@ -13,12 +13,21 @@ The standard [AWS CLI](https://docs.aws.amazon.com/cli/latest/) may also be used
 ```
 aws --endpoint-url http://localhost:8000 s3api list-objects --bucket=mybucket --allow-unordered
 ```
+- Note that the following lines should be added to .aws/credentials:
```
Comment (yuvalif) [inline examples/boto3/README.md:16]:
```suggestion
- Please use the following command to set SNS signature to s3v2:
```
aws configure set default.sns.signature_version s3
```

File: examples/boto3/README.md [hunk 7]
```diff
@@ -13,12 +13,21 @@ The standard [AWS CLI](https://docs.aws.amazon.com/cli/latest/) may also be used
...
+```
```
Comment (yuvalif) [inline examples/boto3/README.md:17]: please remove lines 18-20

**Comment by yuvalif** [comment]
can you please paste sample outputs for the 2 commands you added (creating the topic and getting its attributes)?

File: examples/boto3/README.md [hunk 8]
```diff
@@ -13,12 +13,21 @@ The standard [AWS CLI](https://docs.aws.amazon.com/cli/latest/) may also be used
...
 - Bucket notifications with filtering extensions:
 ```
 aws --region=default --endpoint-url http://localhost:8000 s3api put-bucket-notification-configuration --bucket mybucket --notification-configuration='{"TopicConfigurations": [{"Id": "notif1", "TopicArn": "arn:aws:sns:default::mytopic",  
 "Events": ["s3:ObjectCreated:*", "s3:ObjectRemoved:*"],  
 "Filter": {"Metadata": {"FilterRules": [{"Name": "x-amz-meta-foo", "Value": "bar"}, {"Name": "x-amz-meta-hello", "Value": "world"}]}, "Key": {"FilterRules": [{"Name": "regex", "Value": "([a-z]+)"}]}}}]}'
  ```
+- Topic creation with endpoint:
+```
+aws --endpoint-url http://localhost:8000 sns create-topic --name myname --attributes {'push-endpoint': 'amqp://127.0.0.1:5672', 'amqp-exchange': 'ex1', 'amqp-ack-level': 'broker'}
```
Comment (yuvalif) [inline examples/boto3/README.md:28]:
```suggestion
aws --endpoint-url http://localhost:8000 sns create-topic --name myname --attributes='{"push-endpoint": "amqp://127.0.0.1:5672", "amqp-exchange": "ex1", "amqp-ack-level": "broker"}'
```

File: examples/boto3/topic_attributes.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+if len(sys.argv) == 2:
+    # topic arn as first argument
+    topic_arn = sys.argv[1]
+else:
+    print ('Usage: ' + sys.argv[0] + ' <topic arn> [region name]')
+    sys.exit(1)
+
+# endpoint and keys from vstart
+endpoint = 'http://127.0.0.1:8000'
```
Comment (yuvalif) [inline examples/boto3/topic_attributes.py:20]:
```suggestion
endpoint = '127.0.0.1:8000'
```

File: examples/boto3/topic_attributes.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
+import sys
+import ssl
```
Comment (yuvalif) [inline examples/boto3/topic_attributes.py:2]: no need to support https. please remove

File: examples/boto3/topic_attributes.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+endpoint = 'http://127.0.0.1:8000'
+access_key='0555b35654ad1656d804'
+secret_key='h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q=='
+
+
+parameters = {'Action': 'GetTopic', 'TopicArn': topic_arn}
+body = urlparse.urlencode(parameters)
+string_date = strftime("%a, %d %b %Y %H:%M:%S +0000", gmtime())
+content_type = 'application/x-www-form-urlencoded; charset=utf-8'
+resource = ''
```
Comment (yuvalif) [inline examples/boto3/topic_attributes.py:29]:
```suggestion
resource = '/'
```

File: examples/boto3/topic_attributes.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+method = 'POST'
+string_to_sign = method + '\n\n' + content_type + '\n' + string_date + '\n' + resource 
+signature = base64.b64encode(hmac.new(secret_key.encode('utf-8'), string_to_sign.encode('utf-8'), hashlib.sha1).digest()).decode('ascii')
+headers = {'Authorization': 'AWS '+access_key+':'+signature,
+                   'Date': string_date,
+                   'Host': endpoint,
+                   'Content-Type': content_type}
+if ('https' in endpoint):
```
Comment (yuvalif) [inline examples/boto3/topic_attributes.py:37]: no need to support https. please remove lines 37-39

**Comment by dorindabassey** [comment]
my output for creating topic with endpoint:

dorinda@dorinda-I:~$ aws --endpoint-url https://sns.us-east-1.amazonaws.com/ sns create-topic --name=myname --attributes '{"push-endpoint": "amqp://127.0.0.1:5672", "amqp-exchange": "ex1"}'

An error occurred (InvalidParameter) when calling the CreateTopic operation: Invalid parameter: Attributes Reason: Unknown attribute amqp-exchange

**Comment by yuvalif** [comment]
> my output for creating topic with endpoint:
> 
> dorinda@dorinda-I:~$ aws --endpoint-url https://sns.us-east-1.amazonaws.com/ sns create-topic --name=myname --attributes '{"push-endpoint": "amqp://127.0.0.1:5672", "amqp-exchange": "ex1"}'
> 
> An error occurred (InvalidParameter) when calling the CreateTopic operation: Invalid parameter: Attributes Reason: Unknown attribute amqp-exchange

this makes sense, since AWS don't support our endpoints (e.g .amqp). you should tests the command against a radosgw (you can do that by running a vstart cluster). assuming you are running vstart on your laptop, you should try:
```
aws --endpoint-url http://localhost:8000 sns create-topic --name=myname --attributes '{"push-endpoint": "amqp://127.0.0.1:5672", "amqp-exchange": "ex1"}'
```

**Review by yuvalif — APPROVED** [review]

**Pull request merged**